sigstore/_verify.py

```diff
@@ -0,0 +1,136 @@
+"""
+API for verifying artifact signatures.
+"""
+
+import base64
+import hashlib
+from pathlib import Path
+
+from cryptography.hazmat.primitives import hashes, serialization
+from cryptography.hazmat.primitives.asymmetric import ec
+from cryptography.x509 import (
+    ExtendedKeyUsage,
+    KeyUsage,
+    RFC822Name,
+    SubjectAlternativeName,
+    load_pem_x509_certificate,
+)
+from cryptography.x509.oid import ExtendedKeyUsageOID
+from OpenSSL.crypto import X509, X509Store, X509StoreContext
+
+from sigstore._internal.merkle import verify_merkle_inclusion
+from sigstore._internal.rekor import (
+    RekorClient,
+    RekorEntry,
+    RekorInclusionProof,
+)
+from sigstore._internal.set import verify_set
+
+
+# TODO(alex): Share this with `sign`
+def _no_output(*a, **kw):
+    pass
+
+
+FULCIO_ROOT_CERT = "fulcio.crt.pem"
+
+
+def verify(
+    filename, certificate_path, signature_path, cert_email, output=_no_output
+) -> None:
+    # Read the contents of the package to be verified
+    artifact_contents = filename.read().encode()
+    sha256_artifact_hash = hashlib.sha256(artifact_contents).hexdigest()
+
+    # Load the signing certificate
+    pem_data = certificate_path.read()
+    cert = load_pem_x509_certificate(pem_data)
+
+    # Load the signature
+    b64_artifact_signature = signature_path.read().encode()
+    artifact_signature = base64.b64decode(b64_artifact_signature)
+
+    # In order to verify an artifact, we need to achieve the following:
+    #
+    # 1) Verify that the signing certificate is signed by the root certificate and that the signing
+    #    certificate was valid at the time of signing.
+    # 2) Verify that the signing certiticate belongs to the signer
+    # 3) Verify that the signature was signed by the public key in the signing certificate
+    #
+    # And optionally, if we're performing verification online:
+    #
+    # 4) Verify the inclusion proof supplied by Rekor for this artifact
+    # 5) Verify the Signed Entry Timestamp (SET) supplied by Rekor for this artifact
+
+    # 1) Verify that the signing certificate is signed by the root certificate and that the signing
+    #    certificate was valid at the time of signing.
+    root_cert_path = Path(__file__).parent / FULCIO_ROOT_CERT
+    if not root_cert_path.is_file():
+        # Error
+        return
+    pem_bytes = open(root_cert_path, "rb").read()
+    root = load_pem_x509_certificate(pem_bytes)
+
+    sign_date = cert.not_valid_before
```

Comment:
This chunk leading up to the Rekor interaction is based off the example code in the linked `securesystemslib` issue.